### PR TITLE
Add barrows prayer drain timer

### PR DESCRIPTION
**Pull request by rherms**
Fixes https://github.com/runelite/runelite/issues/2134

This adds a timer that has a duration of 18.2 seconds and only appears inside Barrows crypts or tunnels. I've tested all situations of being in a crypt and then going into a tunnel, being in a crypt/tunnel and logging out, and dying in Barrows. It's my first PR so please point out any mistakes I made! Really hoping this can be in the next release. I've found it quite useful :)

238f546d2eb62ddd370a2ffca5aaf0a1

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPrayerDrainTimer.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import java.time.temporal.ChronoUnit;
+
+import net.runelite.api.SpriteID;
+import net.runelite.client.game.SpriteManager;
+import net.runelite.client.ui.overlay.infobox.InfoBoxPriority;
+import net.runelite.client.ui.overlay.infobox.Timer;
+
+class BarrowsPrayerDrainTimer extends Timer
+{
+    private static final Duration PRAYER_DRAIN_INTERVAL = Duration.of(18200, ChronoUnit.MILLIS);
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPrayerDrainTimer.java]: This can be a long

File: runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPrayerDrainTimer.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2017, Adam <Adam@sigterm.info>
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPrayerDrainTimer.java]: Update copyright

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/infobox/Timer.java [hunk 3]
```diff
@@ -53,6 +53,12 @@ public Timer(long period, ChronoUnit unit, BufferedImage image, Plugin plugin)
 		endTime = startTime.plus(duration);
 	}
 
+	public void reset()
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/infobox/Timer.java]: If you change the fields to `protected` access level here, you can move this method into BarrowsPrayerDrainTimer as a private instance method instead of adding it here.

Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/infobox/Timer.java]: I'll leave my comment from discord here. I'm not sure time should be mutable like this but I don't have a strong opinion.

File: runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java [hunk 4]
```diff
@@ -193,6 +212,10 @@ public void onGameObjectSpawned(GameObjectSpawned event)
 		{
 			ladders.add(gameObject);
 		}
+		if (BARROWS_SARCOPHAGI.contains(gameObject.getId()))
+		{
+			isSarcophagusPresent = true;
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java]: This seems needlessly complex, since all of the sarcophagi are in one region, you can just check the current region

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/infobox/Timer.java (hunk 3 above)
Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/infobox/Timer.java]: I would much prefer just recreating a new timer.

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/infobox/Timer.java]: Or just dont use timer at all, because it do not fits this scenario

File: runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java (hunk 4 above)
Comment (rherms) [reply runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java]: What do you mean by region? Like checking the plane of a player?

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/infobox/Timer.java (hunk 3 above)
Comment (rherms) [reply runelite-client/src/main/java/net/runelite/client/ui/overlay/infobox/Timer.java]: How would you avoid a concurrent modification exception? How could you hook into the timer expiration to create a new one?

File: runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java (hunk 4 above)
Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java]: If you get the worldloc of the player and getRegionID() there is one specific to barrows. It is a 64x64 map square.

File: runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java [hunk 5]
```diff
@@ -220,11 +235,35 @@ public void onGameStateChanged(GameStateChanged event)
...
+		{
+			if (wasInCrypt && !isInCrypt())
+			{
+				prayerDrainTimer.stop();
+			}
+			else if (!wasInCrypt && isInCrypt())
+			{
+				if (prayerDrainTimer == null)
+				{
+					prayerDrainTimer = new BarrowsPrayerDrainTimer(this, spriteManager);
```
Comment (rherms) [inline runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java]: Is there another place I can initialize this so I don't have to mark it as Nullable and lazily initialize it? I tried to do it in startUp, but that resulted in the spriteManager being unable to get the image.

**Comment by rherms** [comment]
Implemented feedback from talking to deathbeam on Discord. This is looking a lot better now!

File: runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java [hunk 6]
```diff
@@ -140,6 +154,7 @@ protected void shutDown()
 		walls.clear();
 		ladders.clear();
 		puzzleAnswer = null;
+		stopPrayerDrainTimer();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java]: you should also reset wasInCrypt to false on plugin shutdown

File: runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java [hunk 7]
```diff
@@ -271,4 +308,28 @@ else if (event.getGroupId() == WidgetID.BARROWS_PUZZLE_GROUP_ID)
 			}
 		}
 	}
+
+	@Subscribe
+	public void onLocalPlayerDeath(LocalPlayerDeath event)
+	{
+		stopPrayerDrainTimer();
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java]: There are more ways to leave than just dieing, and i think this would trigger the gamestate change from above, so why is this here?

File: runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java [hunk 8]
```diff
@@ -155,6 +171,15 @@ protected void shutDown()
 		}
 	}
 
+	@Subscribe
+	public void onConfigChanged(ConfigChanged event)
+	{
+		if (!config.showPrayerDrainTimer())
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java]: This should be checking the config group at least which changed.

File: runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java [hunk 9]
```diff
@@ -271,4 +308,28 @@ else if (event.getGroupId() == WidgetID.BARROWS_PUZZLE_GROUP_ID)
...
+	{
+		if (config.showPrayerDrainTimer())
+		{
+			infoBoxManager.addInfoBox(new BarrowsPrayerDrainTimer(this, spriteManager));
+		}
+	}
+
+	private void stopPrayerDrainTimer()
+	{
+		infoBoxManager.removeIf(t -> t instanceof BarrowsPrayerDrainTimer);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java]:
```suggestion
		infoBoxManager.removeIf(BarrowsPrayerDrainTimer.class::isInstance);
```

File: runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java (hunk 7 above)
Comment (rherms) [reply runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java]: Oh, if a a gamestate change happens on death, then I probably don't need this.

Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java]: I think it does. Why did you add it? You should be testing whether or not the plugin behaves correctly when leaving or teleporting from the crypt.

Comment (rherms) [reply runelite-client/src/main/java/net/runelite/client/plugins/barrows/BarrowsPlugin.java]: I think I just added it because I saw it in another plugin. But you're right, the existing game state change logic catches all those cases.

**Comment by Adam-** [comment]
@deathbeam did you want to wait on #8800 for this?

**Comment by deathbeam** [comment]
No need to wait for it imo, I can always just update my PR after this will be merged.
